django/contrib/admin/views/autocomplete.py

```diff
@@ -41,11 +39,63 @@ def get_paginator(self, *args, **kwargs):
     def get_queryset(self):
         """Return queryset based on ModelAdmin.get_search_results()."""
         qs = self.model_admin.get_queryset(self.request)
+        qs = qs.complex_filter(self.source_field.get_limit_choices_to())
         qs, search_use_distinct = self.model_admin.get_search_results(self.request, qs, self.term)
         if search_use_distinct:
             qs = qs.distinct()
         return qs
 
+    def process_request(self, request):
+        """
+        Validate request integrity, extract and return request parameters.
+
+        Raises:
+            PermissionDenied: If the request is malicious or malformed.
+            Http404: If the target model admin is not configured properly.
+
+        """
+        # extract request parameters
+        term = request.GET.get('term', '')
+        try:
+            app_label = request.GET['app_label']
+            model_name = request.GET['model_name']
+            field_name = request.GET['field_name']
+        except KeyError as e:
+            raise PermissionDenied from e
+
+        # retrieve Django objects from parameters
```

Comment:
"Django" doesn't add much:
```suggestion
        # Retrieve objects from parameters.
```